File: implement-shell-tools/wc/my-wc.js
#!/usr/bin/env node

import { program } from "commander";
import { promises as fs } from "node:fs";

program
  .option("-l", "lines")
  .option("-w", "words")
  .option("-c", "chars")
  .argument("<files...>");

program.parse();

const options = program.opts();
const files = program.args;

let totalL = 0;
let totalW = 0;
let totalC = 0;

for (const file of files) {
  try {
    const content = await fs.readFile(file, "utf-8");

    const lines = content.split("\n").length - 1;
    const words = content.trim() ? content.trim().split(/\s+/).length : 0;
    const chars = content.length;

    totalL += lines;
    totalW += words;
    totalC += chars;

    let result = "";

    if (!options.l && !options.w && !options.c) {
      result = `${lines} ${words} ${chars}`;
    } else {
      if (options.l) result += `${lines} `;
      if (options.w) result += `${words} `;
      if (options.c) result += `${chars} `;
    }

    console.log(result + file);

  } catch (err) {
    console.log(`wc: ${file}: error`);
  }
}

// total если несколько файлов
if (files.length > 1) {
  let result = "";

  if (!options.l && !options.w && !options.c) {
    result = `${totalL} ${totalW} ${totalC}`;
  } else {
    if (options.l) result += `${totalL} `;
    if (options.w) result += `${totalW} `;
    if (options.c) result += `${totalC} `;
  }

  console.log(result + "total");
}

File: implement-shell-tools/wc/my-wc.py
import sys

args = sys.argv[1:]

flag_l = "-l" in args
flag_w = "-w" in args
flag_c = "-c" in args

files = [arg for arg in args if arg not in ["-l", "-w", "-c"]]

if not flag_l and not flag_w and not flag_c:
    flag_l = flag_w = flag_c = True

total_l = total_w = total_c = 0

for file in files:
    try:
        with open(file, "r") as f:
            content = f.read()
    except:
        print(f"wc: {file}: No such file")
        continue

    lines = content.count("\n")
    words = len(content.split()) if content.strip() else 0
    bytes_ = len(content.encode())

    total_l += lines
    total_w += words
    total_c += bytes_

    output = ""
    if flag_l:
        output += f"{lines} "
    if flag_w:
        output += f"{words} "
    if flag_c:
        output += f"{bytes_} "

    print(output + file)

if len(files) > 1:
    total_output = ""
    if flag_l:
        total_output += f"{total_l} "
    if flag_w:
        total_output += f"{total_w} "
    if flag_c:
        total_output += f"{total_c} "

    print(total_output + "total")